### PR TITLE
#167289579 save article as draft

### PR DESCRIPTION
**Pull request by alainburindi**
### What does this PR do?
Have the ``` article's drafting``` working
##### Description of Task to be completed?
* Allow user to save an article as a draft
* Allow user to publish an article
* Allow user to view his drafted articles
* ### [use case](https://docs.google.com/document/d/14b22ao9DW5c1yj-UvPPHEaVIJ9gmt4APapJwEMCmFZA/edit)
##### Have the following endpoints working  
* POST  ```/api/articles```: Create the article as a draft
* GET ```/api/articles/drafts```: Get user's drafted articles
* PATCH ```/api/articles/:slug/publish``` : Publish the article
##### How should this be manually tested?
* Clone the repos
* Install dependencies ```npm install```
* Run ```npm run dev```
* Create an article by pasing at least the title
* Update the created article update route
* Publish the article
##### What are the relevant pivotal tracker stories?
[#167289579](https://www.pivotaltracker.com/story/show/167289579)

### CODE REVIEW

File: src/api/controllers/articleController.js [hunk 1]
```diff
@@ -10,7 +11,7 @@ const {
   Article, Category, User, Report, Share, Highlight
 } = models;
 
-const { CREATED, BAD_REQUEST } = statusCodes;
+const { CREATED, BAD_REQUEST, OK } = statusCodes;
 
 /**
  * containing all article's model controllers
```
Comment (truestbyheart) [inline src/api/controllers/articleController.js]: ``contains all article model`` is more apropriate here

File: src/api/controllers/articleController.js [hunk 2]
```diff
@@ -53,6 +52,110 @@ export default class ArticleController {
     }
   }
 
+  /**
+   * publish a drafted article
```
Comment (truestbyheart) [inline src/api/controllers/articleController.js]: Are we publishing or storing the draft? @alainburindi

File: src/api/controllers/articleController.js [hunk 3]
```diff
@@ -53,6 +52,110 @@ export default class ArticleController {
...
+   * @static
+   * @param {Object} req the request
+   * @param {Object} res the response to be sent
+   * @memberof ArticleController
+   * @returns {Object} res
+   */
+  static async publish(req, res) {
```
Comment (truestbyheart) [inline src/api/controllers/articleController.js:97]: According to the previous comment, I think this name should change... We publish articles that are ready to be viewed by the user and store draft until there ready to be published @alainburindi Please correct me if this is not the right flow of ideas.

**Review by truestbyheart — CHANGES_REQUESTED** [review]
Please I would like more elaboration on the task completed @alainburindi

File: src/api/controllers/articleController.js [hunk 4]
```diff
@@ -53,6 +52,110 @@ export default class ArticleController {
...
+    const { slug } = req.params;
+    const article = await Article.update({
+      status: 'published'
+    },
+    {
```
Comment (premlogic) [inline src/api/controllers/articleController.js:102]: @alainburindi I know that all these attributes have a `not Null ` constraint.
Why would you want to check this yet an article will never be stored if they are null ?

File: src/api/controllers/articleController.js [hunk 5]
```diff
@@ -53,6 +52,110 @@ export default class ArticleController {
...
+        },
+        category: {
+          [Sequelize.Op.ne]: null,
+        },
+        coverImage: {
+          [Sequelize.Op.ne]: null,
+        },
+      }
+    });
+    const published = 1;
```
Comment (premlogic) [inline src/api/controllers/articleController.js:122]: @alainburindi what do you think about `publishedState` ?

File: src/api/controllers/articleController.js [hunk 6]
```diff
@@ -53,6 +52,110 @@ export default class ArticleController {
...
+   */
+  static async getDraftedArticles(req, res) {
+    const defaultLimit = 10;
+    const defaultOffset = 0;
+    const result = {};
+    const emptyArticleArray = 0;
+
+    const limit = req.query.limit || defaultLimit;
+    const offset = req.query.offset || defaultOffset;
+    const articleAttributes = [
```
Comment (premlogic) [inline src/api/controllers/articleController.js]: why do you need to declare these attributes if you know they are going to be fetched exactly as you declared them @alainburindi ?

File: src/api/models/article.js [hunk 7]
```diff
@@ -15,10 +15,10 @@ export default (sequelize, DataTypes) => {
         type: DataTypes.INTEGER
       },
       title: { type: DataTypes.STRING, allowNull: false },
-      description: { type: DataTypes.TEXT, allowNull: false },
-      body: { type: DataTypes.TEXT, allowNull: false },
+      description: { type: DataTypes.TEXT},
```
Comment (premlogic) [inline src/api/models/article.js:17]: @alainburindi why do you need to allow null here ?

File: src/helpers/validators/articleValidator.js [hunk 8]
```diff
@@ -7,18 +7,12 @@ const schema = joi.object().keys({
     .error(() => errorMessage.title)
     .required(),
   description: joi
-    .string()
-    .required()
-    .error(() => errorMessage.description),
+    .string(),
```
Comment (premlogic) [inline src/helpers/validators/articleValidator.js]: I do understand that you removed the required because you allowed null these fields, would you mind keeping the error message as it was @alainburindi ?

File: src/api/controllers/articleController.js (hunk 3 above)
Comment (alainburindi) [reply src/api/controllers/articleController.js:97]: as you can see, this method changes the status of an article to published

File: tests/draft.test.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      .set('Authorization', authToken)
+      .end((err, res) => {
+        res.should.have.status(OK);
+        res.body.should.have.property('message',
+          'congratulations for making it, now you article is published');
+        done();
+      });
+  });
+
+  it('should not publish an article with missing property', (done) => {
```
Comment (premlogic) [inline tests/draft.test.js:40]: Is this the reason why you allowed null @alainburindi ?

**Review by nkpremices — COMMENTED** [review]
Good job @alainburindi.
I left some comment. 
Let's just discuss them then I shall approve

File: tests/draft.test.js (hunk 9 above)
Comment (alainburindi) [reply tests/draft.test.js:40]: yes bro

File: src/helpers/validators/articleValidator.js (hunk 8 above)
Comment (alainburindi) [reply src/helpers/validators/articleValidator.js]: I was having issues to test them but now it's done

File: src/api/controllers/articleController.js (hunk 6 above)
Comment (alainburindi) [reply src/api/controllers/articleController.js]: I didn't see it bro

File: src/api/controllers/articleController.js (hunk 5 above)
Comment (alainburindi) [reply src/api/controllers/articleController.js:122]: is just a number which represents if the article has been published or not

File: src/api/controllers/articleController.js (hunk 2 above)
Comment (alainburindi) [reply src/api/controllers/articleController.js]: we are publishing

File: src/api/controllers/articleController.js (hunk 4 above)
Comment (alainburindi) [reply src/api/controllers/articleController.js:102]: so that an article can not be published, if there is an empty field

**Review by truestbyheart — APPROVED** [review]

**Review by nkpremices — APPROVED** [review]

File: src/api/controllers/articleController.js [hunk 10]
```diff
@@ -41,15 +57,13 @@ export default class ArticleController {
     let articleTags;
 
     const articleValidInput = await articleValidator(req.body);
-    const { category: id } = req.body;
     const readTime = generateReadtime(req.body.body);
+    const { category } = req.body;
```
Comment (MemunaHaruna) [inline src/api/controllers/articleController.js:61]: @alainburindi why this change?

File: src/api/controllers/articleController.js [hunk 11]
```diff
@@ -70,6 +86,84 @@ export default class ArticleController {
...
+        body: {
+          [Sequelize.Op.ne]: null,
+        },
+        description: {
+          [Sequelize.Op.ne]: null,
+        },
+        category: {
+          [Sequelize.Op.ne]: null,
+        },
+        coverImage: {
```
Comment (MemunaHaruna) [inline src/api/controllers/articleController.js:117]: @alainburindi is it compulsory to add a cover image to an article?

File: src/api/controllers/articleController.js [hunk 12]
```diff
@@ -70,6 +86,84 @@ export default class ArticleController {
...
+          [Sequelize.Op.ne]: null,
+        },
+      }
+    });
+    const published = 1;
+    if (article[0] === published) {
+      res.status(OK).json({
+        message: 'congratulations for making it, now you article is published',
```
Comment (MemunaHaruna) [inline src/api/controllers/articleController.js]: @alainburindi change this to: `Your article has been published successfully.`

File: src/api/migrations/20190622195902-create-article.js [hunk 13]
```diff
@@ -7,10 +7,10 @@ export default {
       type: Sequelize.INTEGER
     },
     title: { type: Sequelize.STRING, allowNull: false },
-    description: { type: Sequelize.TEXT, allowNull: false },
```
Comment (MemunaHaruna) [inline src/api/migrations/20190622195902-create-article.js:10]: @alainburindi why remove these checks?

File: src/api/controllers/articleController.js (hunk 3 above)
Comment (MemunaHaruna) [reply src/api/controllers/articleController.js:97]: @alainburindi users should have the option of either publishing an article immediately or saving as a draft first.

Forcing them to store all articles as drafts then later go through another step to publish isn't great UX.

File: src/api/models/article.js (hunk 7 above)
Comment (MemunaHaruna) [reply src/api/models/article.js:17]: @alainburindi can you reply to this question here? Thanks

File: src/api/routes/articleRouter.js [hunk 14]
```diff
@@ -24,6 +24,7 @@ import validateTagsTableQueryRoute from '../../middlewares/validations/query.tag
 const articleRouter = new Router();
 
 articleRouter.get('/search/', articleController.search);
+articleRouter.get('/draft', checkValidToken, articleController.getDraftedArticles);
```
Comment (MemunaHaruna) [inline src/api/routes/articleRouter.js]: @alainburindi `drafts`

File: src/helpers/constants/error.messages.js [hunk 15]
```diff
@@ -14,9 +14,9 @@ export default {
   serverError: 'Something went wrong',
   articleCreate: 'Article created',
   title: 'The title is required',
-  description: 'description is required.',
```
Comment (MemunaHaruna) [inline src/helpers/constants/error.messages.js:17]: @alainburindi please explain these changes.

**Review by MemunaHaruna — CHANGES_REQUESTED** [review]
@alainburindi kindly address all feedback.

File: src/api/controllers/articleController.js (hunk 10 above)
Comment (alainburindi) [reply src/api/controllers/articleController.js]: while creating it! because it is just a draft

Comment (alainburindi) [reply src/api/controllers/articleController.js:61]: because the category can be empty there is no need to check it the database

File: src/api/controllers/articleController.js (hunk 3 above)
Comment (alainburindi) [reply src/api/controllers/articleController.js:97]: from the UI point of view, it will be saved automatically. like on medium

File: src/api/controllers/articleController.js (hunk 11 above)
Comment (alainburindi) [reply src/api/controllers/articleController.js:117]: yes because it is a draft. but before publishing it is required

File: src/api/controllers/articleController.js (hunk 12 above)
Comment (alainburindi) [reply src/api/controllers/articleController.js]: done

File: src/api/routes/articleRouter.js (hunk 14 above)
Comment (alainburindi) [reply src/api/routes/articleRouter.js]: done

File: src/helpers/constants/error.messages.js (hunk 15 above)
Comment (alainburindi) [reply src/helpers/constants/error.messages.js:17]: because while saving as draft these fields can be empty

File: src/api/controllers/articleController.js [hunk 16]
```diff
@@ -41,15 +57,13 @@ export default class ArticleController {
...
     const articleValidInput = await articleValidator(req.body);
-    const { category: id } = req.body;
     const readTime = generateReadtime(req.body.body);
+    const { category } = req.body;
     let coverImage;
     if (req.file) {
       const savedFile = await cloudinary.v2.uploader.upload(req.file.path);
       coverImage = savedFile.secure_url;
     }
-    const getCategory = await Category.findOne({ where: { id } });
```
Comment (MemunaHaruna) [inline src/api/controllers/articleController.js:51]: @alainburindi please add some context as to why you've made this change.

**Review by alainburindi — COMMENTED** [review]
All changes I made about the model help to save an article as a draft which means that some fields allow to be null

**Review by MemunaHaruna — APPROVED** [review]
